src/Microsoft.OData.Client/BulkUpdateGraph.cs

```diff
@@ -0,0 +1,175 @@
+                {
+                    throw Error.InvalidOperation(Strings.DataBinding_Util_UnknownEntitySetName(parentDescriptor.Entity.GetType().FullName));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Links a descriptor with its related descriptors.
+        /// </summary>
+        /// <param name="parent">The parent descriptor.</param>
+        /// <param name="child">The related descriptor to the parent descriptor.</param>
+        public void AddRelatedDescriptor(Descriptor parent, Descriptor child)
+        {
+            HashSet<Descriptor> childrenDescriptors = this.GetRelatedDescriptors(parent);
+
+            if (childrenDescriptors.Count > 0)
+            {
+                // If the parent already has some related descriptors then we don't need to 
+                // create a new parent-child entry in this.relatedDescriptors dictionary, we just add the child descriptor to the
+                // parent descriptors related collection. 
+                childrenDescriptors.Add(child);
+            }
+            else 
+            {
+                childrenDescriptors.Add(child);
+            }      
+        }
+
+        /// <summary>
+        /// Adds to the top-level descriptors' list.
+        /// </summary>
+        /// <param name="descriptor">The descriptor to be added to the top-level descriptors' list.</param>
+        public void AddTopLevelDescriptors(EntityDescriptor descriptor)
+        {
+            if (this.topLevelDescriptors == null)
+            {
+                this.topLevelDescriptors = new List<Descriptor>();
+            }
+
+            this.topLevelDescriptors.Add(descriptor);
+        }
+
+        /// <summary>
+        /// Gets the descriptors of the entities related to the entity associated with the provided descriptor
+        /// </summary>
+        /// <param name="descriptor">The descriptor.</param>
+        /// <returns>All the related descriptors to a given key descriptor.</returns>
+        public HashSet<Descriptor> GetRelatedDescriptors(Descriptor descriptor)
+        {
+            if (this.descriptorGraph.TryGetValue(descriptor, out HashSet<Descriptor> relatedDescriptors))
+            {
+                return relatedDescriptors;
+            }
+
+            relatedDescriptors = new HashSet<Descriptor>();
+            this.descriptorGraph[descriptor] = relatedDescriptors;
+
+            return relatedDescriptors;
+        }
+
+        /// <summary>
+        /// This method checks whether the descriptor has already been added to the 
+        /// graph to avoid cyclic operations. If the descriptor has already been added, we don't do the recursive call. 
+        /// </summary>
+        /// <param name="descriptor">The descriptor to check if it already exists in the graph.</param>
+        /// <returns>true if the link already exists, otherwise returns false.</returns>
+        public bool CheckIfLinkAlreadyExists(Descriptor descriptor)

```

Comment:
Suggesting a simpler name.
```suggestion
        public bool HasDescriptor(Descriptor descriptor)

```
or even simpler:
```suggestion
        public bool Contains(Descriptor descriptor)

```

Also reads better in if statements:
`if (graph.Contains(descriptor))` vs `if (graph.CheckIfLinkAlreadyExists(descriptor))`